File: packages/langchain/langchain_oneshot/tools/commerce.py
"""OneShot commerce tools."""

from __future__ import annotations

import json
from typing import Any, Type

from langchain_core.tools import BaseTool
from pydantic import BaseModel

from langchain_oneshot._types import CommerceBuyInput, CommerceSearchInput
from oneshot import OneShotClient


def _strip_none(d: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in d.items() if v is not None}


class CommerceSearchTool(BaseTool):
    """Search for products to purchase. Free to search."""

    name: str = "oneshot_commerce_search"
    description: str = (
        "Search for products available for purchase. Provide a search query. "
        "Free to search — you only pay when buying."
    )
    args_schema: Type[BaseModel] = CommerceSearchInput
    handle_tool_error: bool = True

    client: OneShotClient

    def _run(self, **kwargs: Any) -> str:
        result = self.client.call_tool("/v1/tools/commerce/search", _strip_none(kwargs))
        return json.dumps(result)

    async def _arun(self, **kwargs: Any) -> str:
        result = await self.client.acall_tool(
            "/v1/tools/commerce/search", _strip_none(kwargs)
        )
        return json.dumps(result)


class CommerceBuyTool(BaseTool):
    """Purchase a product via OneShot. Product price + service fee."""

    name: str = "oneshot_commerce_buy"
    description: str = (
        "Purchase a product. Requires a product_url and a shipping_address "
        "with first_name, last_name, street, city, state, zip_code, and phone. "
        "Cost is product price plus a service fee."
    )
    args_schema: Type[BaseModel] = CommerceBuyInput
    handle_tool_error: bool = True

    client: OneShotClient

    def _run(self, **kwargs: Any) -> str:
        payload = _prepare_buy_payload(kwargs)
        result = self.client.call_tool(
            "/v1/tools/commerce/buy", payload, timeout_sec=180
        )
        return json.dumps(result)

    async def _arun(self, **kwargs: Any) -> str:
        payload = _prepare_buy_payload(kwargs)
        result = await self.client.acall_tool(
            "/v1/tools/commerce/buy", payload, timeout_sec=180
        )
        return json.dumps(result)


def _prepare_buy_payload(kwargs: dict[str, Any]) -> dict[str, Any]:
    """Convert Pydantic-parsed args to the API payload shape."""
    payload: dict[str, Any] = {
        "product_url": kwargs["product_url"],
        "shipping_address": kwargs["shipping_address"],
    }
    if isinstance(payload["shipping_address"], BaseModel):
        payload["shipping_address"] = payload["shipping_address"].model_dump(exclude_none=True)
    if kwargs.get("quantity") is not None:
        payload["quantity"] = kwargs["quantity"]
    if kwargs.get("variant_id") is not None:
        payload["variant_id"] = kwargs["variant_id"]
    return payload

File: packages/mcp-server/src/tools/commerce.ts
import { Tool } from "@modelcontextprotocol/sdk/types.js";
import { OneShot } from "@oneshot-agent/sdk";

export const commerceSearchTool: Tool = {
  name: "oneshot_commerce_search",
  description: "Search for products to purchase. Free to search, pay only when buying.",
  inputSchema: {
    type: "object",
    properties: {
      query: {
        type: "string",
        description: "Product search query",
      },
      limit: {
        type: "number",
        description: "Maximum number of results (default: 10)",
      },
    },
    required: ["query"],
  },
};

export const commerceBuyTool: Tool = {
  name: "oneshot_commerce_buy",
  description: "Purchase a product. Requires product URL and shipping address. Price + service fee.",
  inputSchema: {
    type: "object",
    properties: {
      product_url: {
        type: "string",
        description: "Product URL to purchase",
      },
      shipping_address: {
        type: "object",
        properties: {
          first_name: { type: "string" },
          last_name: { type: "string" },
          street: { type: "string" },
          street2: { type: "string" },
          city: { type: "string" },
          state: { type: "string" },
          zip_code: { type: "string" },
          country: { type: "string" },
          email: { type: "string" },
          phone: { type: "string" },
        },
        required: ["first_name", "last_name", "street", "city", "state", "zip_code", "phone"],
      },
      quantity: {
        type: "number",
        description: "Quantity to purchase (default: 1)",
      },
      variant_id: {
        type: "string",
        description: "Product variant ID (optional)",
      },
    },
    required: ["product_url", "shipping_address"],
  },
};

export async function handleCommerceSearch(agent: OneShot, args: Record<string, unknown>) {
  return agent.commerceSearch({
    query: args.query as string,
    limit: args.limit as number | undefined,
  });
}

export async function handleCommerceBuy(agent: OneShot, args: Record<string, unknown>) {
  const shippingAddress = args.shipping_address as Record<string, string>;
  return agent.commerceBuy({
    product_url: args.product_url as string,
    shipping_address: {
      first_name: shippingAddress.first_name,
      last_name: shippingAddress.last_name,
      street: shippingAddress.street,
      street2: shippingAddress.street2,
      city: shippingAddress.city,
      state: shippingAddress.state,
      zip_code: shippingAddress.zip_code,
      country: shippingAddress.country,
      email: shippingAddress.email,
      phone: shippingAddress.phone,
    },
    quantity: args.quantity as number | undefined,
    variant_id: args.variant_id as string | undefined,
  });
}
